File: android/app/src/main/java/app/lovable/c083a032f07147138f072f3c0dc36e56/PermissionsManager.java
package app.lovable.c083a032f07147138f072f3c0dc36e56;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class PermissionsManager {
    public static final int REQ_MEDIA = 1001;
    public static final int REQ_NOTIFICATIONS = 1002;
    private static final String PREFS = "pipskit_permissions";
    private static final String KEY_ASKED_MEDIA = "asked_media_v2";
    private static final String KEY_ASKED_NOTIFS = "asked_notifs_v1";
    private final Activity activity;
    private final SharedPreferences prefs;

    public PermissionsManager(Activity activity) {
        this.activity = activity;
        this.prefs = activity.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    private String[] mediaPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return new String[] {
                Manifest.permission.READ_MEDIA_IMAGES,
                Manifest.permission.READ_MEDIA_VIDEO
            };
        }
        return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
    }

    public boolean hasNotificationPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return true;
        return ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS)
            == PackageManager.PERMISSION_GRANTED;
    }

    public void maybeRequestNotificationsOnFirstLaunch() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return;
        if (hasNotificationPermission() || prefs.getBoolean(KEY_ASKED_NOTIFS, false)) return;
        prefs.edit().putBoolean(KEY_ASKED_NOTIFS, true).apply();
        ActivityCompat.requestPermissions(activity,
            new String[] { Manifest.permission.POST_NOTIFICATIONS }, REQ_NOTIFICATIONS);
    }

    public boolean hasMediaPermissions() {
        for (String permission : mediaPermissions()) {
            if (ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) return false;
        }
        return true;
    }

    public void maybeRequestMediaOnFirstLaunch() {
        if (hasMediaPermissions() || prefs.getBoolean(KEY_ASKED_MEDIA, false)) return;
        prefs.edit().putBoolean(KEY_ASKED_MEDIA, true).apply();
        ActivityCompat.requestPermissions(activity, mediaPermissions(), REQ_MEDIA);
    }

    public void requestMediaNow() {
        if (hasMediaPermissions()) return;
        String[] permissions = mediaPermissions();
        boolean canShowPrompt = !prefs.getBoolean(KEY_ASKED_MEDIA, false);
        for (String permission : permissions) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) canShowPrompt = true;
        }
        if (canShowPrompt) {
            prefs.edit().putBoolean(KEY_ASKED_MEDIA, true).apply();
            ActivityCompat.requestPermissions(activity, permissions, REQ_MEDIA);
        } else {
            showOpenSettingsDialog();
        }
    }

    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode != REQ_MEDIA) return;
        boolean allGranted = grantResults.length > 0;
        for (int result : grantResults) allGranted = allGranted && result == PackageManager.PERMISSION_GRANTED;
        if (!allGranted) showOpenSettingsDialog();
    }

    private void showOpenSettingsDialog() {
        new AlertDialog.Builder(activity)
            .setTitle(R.string.permission_storage_rationale_title)
            .setMessage(R.string.permission_denied_settings)
            .setPositiveButton(R.string.permission_open_settings, (dialog, which) -> {
                Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setData(Uri.fromParts("package", activity.getPackageName(), null));
                activity.startActivity(intent);
            })
            .setNegativeButton(R.string.permission_cancel, null)
            .show();
    }
}

File: android-app/app/src/main/java/com/smarttradetracker/app/PermissionsManager.kt
package com.smarttradetracker.app

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

/**
 * Native Android permissions manager.
 * - Requests modern granular media permissions (Android 13+)
 * - Falls back to READ_EXTERNAL_STORAGE for older devices
 * - Requests only once on first launch
 * - Shows rationale + open-settings dialog on denial
 */
class PermissionsManager(private val activity: Activity) {

    companion object {
        const val REQ_MEDIA = 1001
        private const val PREFS = "pipskit_permissions"
        private const val KEY_ASKED_MEDIA = "asked_media_v1"
    }

    private val prefs: SharedPreferences =
        activity.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun mediaPermissions(): Array<String> =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            arrayOf(
                Manifest.permission.READ_MEDIA_IMAGES,
                Manifest.permission.READ_MEDIA_VIDEO,
                Manifest.permission.READ_MEDIA_AUDIO
            )
        } else {
            arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        }

    fun hasMediaPermissions(): Boolean = mediaPermissions().all {
        ContextCompat.checkSelfPermission(activity, it) == PackageManager.PERMISSION_GRANTED
    }

    /** Request once on first launch. No-op if already granted or already asked. */
    fun maybeRequestMediaOnFirstLaunch() {
        if (hasMediaPermissions()) return
        if (prefs.getBoolean(KEY_ASKED_MEDIA, false)) return
        prefs.edit().putBoolean(KEY_ASKED_MEDIA, true).apply()
        ActivityCompat.requestPermissions(activity, mediaPermissions(), REQ_MEDIA)
    }

    /** Force a request (called from JS bridge when user retries from web UI). */
    fun requestMediaNow() {
        if (hasMediaPermissions()) return
        val perms = mediaPermissions()
        val canShowSystemPrompt = perms.any {
            ActivityCompat.shouldShowRequestPermissionRationale(activity, it)
        } || !prefs.getBoolean(KEY_ASKED_MEDIA, false)

        if (canShowSystemPrompt) {
            prefs.edit().putBoolean(KEY_ASKED_MEDIA, true).apply()
            ActivityCompat.requestPermissions(activity, perms, REQ_MEDIA)
        } else {
            showOpenSettingsDialog()
        }
    }

    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        if (requestCode != REQ_MEDIA) return
        val allGranted = grantResults.isNotEmpty() &&
            grantResults.all { it == PackageManager.PERMISSION_GRANTED }
        if (!allGranted) {
            val permanentlyDenied = permissions.any {
                !ActivityCompat.shouldShowRequestPermissionRationale(activity, it)
            }
            if (permanentlyDenied) showOpenSettingsDialog() else showRationaleDialog()
        }
    }

    private fun showRationaleDialog() {
        AlertDialog.Builder(activity)
            .setTitle(R.string.permission_storage_rationale_title)
            .setMessage(R.string.permission_storage_rationale_message)
            .setPositiveButton(R.string.retry) { _, _ -> requestMediaNow() }
            .setNegativeButton(R.string.permission_cancel, null)
            .setCancelable(true)
            .show()
    }

    private fun showOpenSettingsDialog() {
        AlertDialog.Builder(activity)
            .setTitle(R.string.permission_storage_rationale_title)
            .setMessage(R.string.permission_denied_settings)
            .setPositiveButton(R.string.permission_open_settings) { _, _ ->
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                    data = Uri.fromParts("package", activity.packageName, null)
                }
                activity.startActivity(intent)
            }
            .setNegativeButton(R.string.permission_cancel, null)
            .show()
    }
}
